File: Python/merge_sort_step.py
import collections
import csv

class TestMergeSortStep:
    def format_output_list(self, data_list):
        """
        Converts each (number, string) pair into a 'number/string' format.
        This helps make the output easier to read when showing the sorting steps.
        """
        return [f"{item[0]}/{item[1]}" for item in data_list]

    def main(self):
        """
        Main function that controls the process:
        - Reads the CSV file
        - Asks the user which rows to sort
        - Performs Merge Sort with step-by-step logging
        - Saves all steps to a text file
        """
        # Ask the user to enter the name of the CSV file (should be in the same folder)
        input_filename = input("Enter CSV file name : ")    
        if not input_filename.endswith('.csv'):
            print("Error: Please provide a valid CSV file.")
            return
        
        # Read all rows from the file so we can extract the part the user wants
        full_data = self.read_data_from_csv(input_filename)
        
        if full_data is None:
            print(f"Error: Could not read data from {input_filename}. Please ensure the file exists and is correctly formatted.")
            return

        # Let the user select a range of rows to sort (e.g., from row 10 to 20)
        while True:
            try:
                start_row_input = input(f"Enter start row : ")
                end_row_input = input(f"Enter end row : ")
                
                # Convert to 0-based index (Python indexing starts from 0)
                start_idx = int(start_row_input) - 1
                end_idx = int(end_row_input) - 1

                # Make sure the selected range is valid
                if not (0 <= start_idx <= end_idx < len(full_data)):
                    print(f"Invalid row range. Please enter values between 1 and {len(full_data)} where start row <= end row.")
                    continue
                break
            except ValueError:
                print("Invalid input. Please enter integers for start and end rows.")
        
        # Get the portion of the data the user wants to sort
        # We make a copy of it to sort, so the original data stays unchanged
        data_subset = full_data[start_idx : end_idx + 1]

        # Generate the output filename based on selected row range
        output_filename = f"merge_sort_step_{start_row_input}_{end_row_input}.txt"
        
        try:
            # Open the output file and start logging the sorting steps
            with open(output_filename, 'w', newline='') as output_file:
                # Write the initial state of the array before any sorting
                output_file.write(f"[{', '.join(self.format_output_list(data_subset))}]\n")
                
                # Perform merge sort and record each step
                self._mergeSort(data_subset, 0, len(data_subset) - 1, output_file)
                
            print(f"Sorting steps written to {output_filename}")
        except Exception as e:
            print(f"An error occurred during sorting or writing output: {e}")

    def read_data_from_csv(self, filename):
        """
        Reads data from a CSV file. Each row should have two values: a number and a string.
        Returns a list of (number, string) pairs.
        If a row is missing data or contains invalid data, it's skipped with a warning.
        """
        data = []
        try:
            with open(filename, 'r', newline='') as csvfile:
                reader = csv.reader(csvfile)
                for row_num, row in enumerate(reader, 1):
                    if len(row) == 2:
                        try:
                            number = int(row[0].strip())  # Convert the first column to an integer
                            label = row[1].strip()        # Trim any extra spaces from the second column
                            data.append((number, label))
                        except ValueError:
                            print(f"Skipping row {row_num} due to non-integer value in first column: {row}")
                    else:
                        print(f"Skipping row {row_num} due to malformed row (expected 2 columns): {row}")
            return data
        except FileNotFoundError:
            print(f"Error: File '{filename}' not found. Please make sure the CSV file is in the same directory.")
            return None
        except Exception as e:
            print(f"An unexpected error occurred while reading the CSV: {e}")
            return None

    def _mergeSort(self, S, left, right, output_file):
        """
        Recursive Merge Sort function that breaks the list into halves,
        sorts them, and then merges them while logging each merge step.
        """
        if left < right:
            mid = (left + right) // 2
            self._mergeSort(S, left, mid, output_file)       # Sort the first half
            self._mergeSort(S, mid + 1, right, output_file)  # Sort the second half
            self.merge(S, left, mid, right, output_file)     # Merge the two halves

    def merge(self, S, left, mid, right, output_file):
        """
        This function merges two sorted parts of the array into one.
        It also logs the full state of the array after each merge.
        """
        # Copy the left part into a deque (so we can pop from the front efficiently)
        L = collections.deque()
        for i in range(mid - left + 1):
            L.append(S[left + i])

        # Copy the right part into another deque
        R = collections.deque()
        for j in range(right - mid):
            R.append(S[mid + 1 + j])

        k = left  # Index for writing merged results back into the array

        # Merge from both deques into the original list
        while L and R:
            if L[0][0] <= R[0][0]:  # Compare based on the integer part
                S[k] = L.popleft()
            else:
                S[k] = R.popleft()
            k += 1

        # If anything is left in L, copy it over
        while L:
            S[k] = L.popleft()
            k += 1

        # If anything is left in R, copy it over
        while R:
            S[k] = R.popleft()
            k += 1
        
        # Log the full array state after each merge
        output_file.write(f"[{', '.join(self.format_output_list(S))}]\n")


if __name__ == "__main__":
    sorter = TestMergeSortStep()
    sorter.main()

File: C++/merge_sort_step.cpp
#include <iostream>     
#include <fstream>     
#include <sstream>      
#include <vector>       
#include <string>       
#include <limits>       

using namespace std;     


class TestMergeSortStep {
public:
    void main() {
        string input_filename;
        cout << "Enter CSV file name : ";          // Ask user to input the CSV filename
        cin >> input_filename;                     // Store input filename

        // Check if file has .csv extension
        if (input_filename.size() < 4 || input_filename.substr(input_filename.size() - 4) != ".csv") {
            cerr << "Error: Please provide a valid CSV file." << endl;
            return;                                
        }

        // Load all data from the CSV file
        vector<pair<int, string>> full_data = read_data_from_csv(input_filename);
        if (full_data.empty()) {
            cerr << "Error: Could not read data from " << input_filename << endl;
            return;                                
        }

        int start_row, end_row;
        while (true) {
            cout << "Enter start row : ";           // Ask user for the starting row
            if (!(cin >> start_row)) {              // Validate input is an integer
                cout << "Invalid input. Please enter integers." << endl;
                cin.clear(); 
                cin.ignore(numeric_limits<streamsize>::max(), '\n'); 
                continue;
            }
            cout << "Enter end row : ";             // Ask user for the ending row
            if (!(cin >> end_row)) {                // Validate input is an integer
                cout << "Invalid input. Please enter integers." << endl;
                cin.clear(); 
                cin.ignore(numeric_limits<streamsize>::max(), '\n'); 
                continue;
            }
            --start_row;                            // Convert to 0-based index
            --end_row;
            if (start_row < 0 || end_row >= full_data.size() || start_row > end_row) {
                cout << "Invalid row range. Please enter values between 1 and " << full_data.size() << ", with start <= end." << endl;
                continue;                           
            }
            break;                               
        }

        // Extract the selected subset of data
        vector<pair<int, string>> data_subset(full_data.begin() + start_row, full_data.begin() + end_row + 1);

        // Create output filename with selected row range
        string output_filename = "merge_sort_step_" + to_string(start_row + 1) + "_" + to_string(end_row + 1) + ".txt";

        // Open the output file to log sorting steps
        ofstream output_file(output_filename);
        if (!output_file.is_open()) {
            cerr << "Error opening output file." << endl;
            return;
        }

        // Write the initial state of the subset to the output file
        output_file << "[" << format_output_list(data_subset) << "]\n";

        // Perform merge sort with step logging
        _mergeSort(data_subset, 0, data_subset.size() - 1, output_file);

        // Notify user sorting is completed
        cout << "Sorting steps written to " << output_filename << endl;
    }

private:
    // Read and parse data from a CSV file into a vector of (int, string) pairs
    vector<pair<int, string>> read_data_from_csv(const string& filename) {
        vector<pair<int, string>> data;
        ifstream file(filename);          // Open file for reading
        string line;

        if (!file.is_open()) {           // Check if file opened successfully
            cerr << "File '" << filename << "' not found." << endl;
            return data;
        }

        int row_num = 0;
        while (getline(file, line)) {    // Read each line
            ++row_num;
            stringstream ss(line);
            string num_str, word;
            if (getline(ss, num_str, ',') && getline(ss, word)) { // Parse number and word
                try {
                    int number = stoi(num_str);      // Convert to integer
                    data.emplace_back(number, word); // Store valid row
                } catch (...) {
                    cerr << "Skipping row " << row_num << " due to non-integer value: " << line << endl;
                }
            } else {
                cerr << "Skipping row " << row_num << " due to malformed row: " << line << endl;
            }
        }

        return data;  
    }

    // Format the vector into a readable string for logging
    string format_output_list(const vector<pair<int, string>>& data_list) {
        stringstream ss;
        for (size_t i = 0; i < data_list.size(); ++i) { //This loop formats the contents of the data_list vector into a readable string.
            ss << data_list[i].first << "/" << data_list[i].second; //each element, it adds the integer and string (separated by a slash) to the stringstream ss.
            if (i < data_list.size() - 1) ss << ", "; //If the element is not the last one, it also adds a comma 
        }
        return ss.str();  
    }

    // Recursive merge sort function with output logging
    void _mergeSort(vector<pair<int, string>>& S, int left, int right, ofstream& output_file) {
        if (left < right) {
            int mid = (left + right) / 2;                  // Find midpoint
            _mergeSort(S, left, mid, output_file);         // Sort left half
            _mergeSort(S, mid + 1, right, output_file);    // Sort right half
            merge(S, left, mid, right, output_file);       // Merge both halves
        }
    }

    // Merge two sorted halves of the array with logging
    void merge(vector<pair<int, string>>& S, int left, int mid, int right, ofstream& output_file) {
        int n1 = mid - left + 1;       // Size of left subarray
        int n2 = right - mid;          // Size of right subarray

        vector<pair<int, string>> L(n1), R(n2);  // Temporary arrays

        // part of the merge step in merge sort.
        // It works by copying elements from the original array S, starting at index left, 
        // into the temporary array L. The loop runs n1 times, 
        // so it copies all elements from S[left] up to S[mid] into L[0] up to L[n1-1].
        for (int i = 0; i < n1; ++i) L[i] = S[left + i];        // Copy left half

        //This line copies elements from the right half of the array S (from index mid + 1 to right) into the temporary array R. 
        // The loop runs n2 times, so it fills R[0] to R[n2-1] with the elements S[mid + 1] to S[right].
        for (int j = 0; j < n2; ++j) R[j] = S[mid + 1 + j];     // Copy right half


        //i is used to track the current position in the left temporary array (L).
        //j is used to track the current position in the right temporary array (R).
        //k = left: index for placing merged elements back into the original array S, starting at the left boundary. 
        // These variables help merge the two sorted halves back into S in the correct order.
        int i = 0, j = 0, k = left;

        // Merge L and R back into S in sorted order
        while (i < n1 && j < n2) {          // Continue while both subarrays have elements left
            if (L[i].first <= R[j].first) { // If current element in left subarray is less than or equal to right
                S[k++] = L[i++];            // Copy element from left subarray to S and move to next in left
            } else {
                S[k++] = R[j++];            // Otherwise, copy element from right subarray to S and move to next in right
            }
        }

        // Copy any remaining elements in L
        while (i < n1) S[k++] = L[i++];
        //This line copies any remaining elements from the left temporary array L into the original array S. 
        // It runs after the main merge loop, in case there are still elements left in L (because all elements in R have already been merged). 
        // For each remaining element, it assigns L[i] to S[k], 
        // then increments both i and k to move to the next positions.

        // Copy any remaining elements in R
        while (j < n2) S[k++] = R[j++];
        //This line copies any remaining elements from the right temporary array R into the original array S. 
        // It runs after the main merge loop, in case there are still elements left in R (because all elements in L have already been merged). 
        // For each remaining element, it assigns R[j] to S[k], 
        // then increments both j and k to move to the next positions.

        // Log current array state after merging
        output_file << "[" << format_output_list(S) << "]\n";
    }
};

int main() {
    TestMergeSortStep sorter;  
    sorter.main();             
    return 0;                  
}
